autoprecompiles/src/memory_optimizer.rs

```diff
@@ -74,76 +78,78 @@ enum MemoryOp {
 }
 
 #[derive(Clone, Debug)]
-struct MemoryBusInteraction<T> {
+struct MemoryBusInteraction<T: FieldElement, V> {
     op: MemoryOp,
     address_space: T,
-    addr: AlgebraicExpression<T>,
-    data: Vec<AlgebraicExpression<T>>,
-}
-
-impl<T: FieldElement> MemoryBusInteraction<T> {
-    fn try_addr_u32(&self) -> Option<u32> {
-        match self.addr {
-            AlgebraicExpression::Number(n) => n.to_integer().try_into_u32(),
-            _ => None,
-        }
-    }
+    addr: QuadraticSymbolicExpression<T, V>,
+    data: Vec<QuadraticSymbolicExpression<T, V>>,
+    #[allow(dead_code)]
+    timestamp: QuadraticSymbolicExpression<T, V>,
 }
 
-impl<T: FieldElement> MemoryBusInteraction<T> {
-    /// Tries to convert a `SymbolicBusInteraction` to a `MemoryBusInteraction`.
+impl<T: FieldElement, V: Ord + Clone + Eq + Display> MemoryBusInteraction<T, V> {
+    /// Tries to convert a `BusInteraction` to a `MemoryBusInteraction`.
     ///
     /// Returns `Ok(None)` if we know that the bus interaction is not a memory bus interaction.
     /// Returns `Err(_)` if the bus interaction is a memory bus interaction but could not be converted properly
     /// (usually because the multiplicity is not -1 or 1).
     /// Otherwise returns `Ok(Some(memory_bus_interaction))`
-    fn try_from_symbolic_bus_interaction(
-        bus_interaction: &SymbolicBusInteraction<T>,
+    fn try_from_bus_interaction(
+        bus_interaction: &BusInteraction<QuadraticSymbolicExpression<T, V>>,
     ) -> Result<Option<Self>, ()> {
-        if bus_interaction.id != MEMORY_BUS_ID {
-            return Ok(None);
+        match bus_interaction.bus_id.try_to_number() {
+            None => return Err(()),
+            Some(id) if id == MEMORY_BUS_ID.into() => {}
+            Some(_) => return Ok(None),
         }
-        // TODO: Timestamp is ignored, we could use it to assert that the bus interactions
-        // are in the right order.
-        let AlgebraicExpression::Number(address_space) = bus_interaction.args[0].clone() else {
-            panic!("Address space must be known!")
-        };
-        let op = match bus_interaction.try_multiplicity_to_number() {
+
+        let op = match bus_interaction.multiplicity.try_to_number() {
             Some(n) if n == 1.into() => MemoryOp::Send,
             Some(n) if n == (-1).into() => MemoryOp::Receive,
             _ => return Err(()),
         };
-        let addr = bus_interaction.args[1].clone();
-        let data = bus_interaction.args[2..bus_interaction.args.len() - 1].to_vec();
+
+        // TODO: Timestamp is ignored, we could use it to assert that the bus interactions
```

Comment:
Ah, because of this

Reply:
I think we should move this comment to the `MemoryBusInteraction::timestamp` field. Then the `#[allow(dead_code)]` becomes clear. Also it's not like we could check the order here, as we're only given one interaction at a time!